### DISCUSSION TITLE
browsing for internet radio stations

### DISCUSSION THREAD

**Discussion by marc-portier** [micro-nova/AmpliPi#209] [Ideas]
testing out the web UI on my ubuntu I came across this screen to configure radio-stations:

![image](https://user-images.githubusercontent.com/299574/140513472-275592a5-2b96-4afe-9a43-4c0e3001f66c.png)

Which assumes people know how to grab the required information through available radio-directory-services?

To lower that threshold  I'd like to sign up for helping to integrate a search-and-browse action towards a radio-directory into that screen.

Have come across https://github.com/andreztz/pyradios which is using the open services from https://api.radio-browser.info/ that seems to be a good starting point (Licenses on both py lib and open service seem to be compatible with this project at first glance).  But open to other suggestions.  (companion project https://github.com/andreztz/xradios adds a CLI experience to that)

Thinking about a layered approach:
- wrapping the general purpose pyradios lib in a tuned-for AmpliPi usage service  (also allowing to replace the current choice of lib and service if ever needed)
- extending the REST API with some paths to access this service
- finally mixin in the browsing feature into the web-ui

Also:  I think I read somewhere that the front-end stack is intended to move to using react-js?  Is that in the works somewhere already?  Near term plans?  Best current UI-dev strategy to apply?

Other remarks or suggestions?

**Comment by linknum23** [comment]
This is a great idea! I think this would really make this easy to use.

My instincts are to add this functionality to just the webapp, skipping to integrating it in with the API, since the primary use case is for this settings panel. It looks like there is a js library here https://github.com/nepodev/radio-browser that is built for interfacing with [https://api.radio-browser.info/](https://api.radio-browser.info/), the examples for that look like a good place to start.

As for the webapp we are planning to upgrade the interface to make it easier to convert it to a lightweight android/ios app. No work has been started though, so for now I would focus on integrating this into the current settings menu.

**Reply by marc-portier** [reply]
Ok. Will investigate into the js lib. Hope it will not fight with cors issues though (which was my thinking behind the API expansion, but I do follow your motivation and the service providers should be the ones doing the actual cors fighting 🧐)

**Comment by rickballard** [comment]
The hard part is finding an URL that autoplays.  Many listed urls, like those in Wikipedia, link to a stream__stopped mode, and you have to then press a button on the page Wikipedia links to before the audio starts.  Meanwhile, the Pandora stream option does not work in too many places, outside of the US, such as here in Canada.  Plex connections can require your Plex Media Server to be running in a Server that is running full-time, which your Invidia Shield, TV Dongle, or HTPC may not do.

**Reply by linknum23** [reply]
For checking urls, I would recommend using VLC. Use the 'Media' -> 'Open Network Stream' -> 'Play' sequence to test urls.  We would love to integrate an internet radio stream selector, but we aren't there yet.

Also would you mind providing an example url that did not work?

**Reply by linknum23** [reply]
As for the feedback on your initial experience, we would love to hear your experience in https://amplipi.discourse.group/c/feedback. For troubleshooting your app issues (likely an MDNS network configuration issue) I would suggest https://amplipi.discourse.group/c/troubleshooting, I'm afraid this discussion is the wrong place.

**Reply by rickballard** [reply]
TEMPORARILY SOLVED (the app will include a repository eventually): 

VLC is good for testing URL's, but it does not help you find local FM radio station streams.  FM radio stations do not typically make it easy to find their streaming-URL's, but fmstream-dot-org lists streaming-URL's for internet radio stations with FM-broadcasts, in many countries.  

Note: The higher bitrate streams may take a few seconds to load each time you turn them on, and oddly enough, the MP3 and playlist-type streams also seem slow to start.  To use the actual off-air FM-broadcast, AmpliPi needs a dongle in the back to convert waves into bits. 

To get an URL for the radio station's logo, go to the radio station's commercial website and (right-click to inspect the entire page source, or to inspect the actual element you just pressed to make the broadcast start -- page code often changes, depending upon whether the broadcast is active or not.  Keep digging through the code, while the stream is still playing. 

You are looking for an image that is quite small.  Yet, when you test the logo's URL in a browser window address bar, the logo image may fill up the entire browser window, no matter how big you make the window.  Some images are generated via an array of discrete values.  In this case, take the easy route.  You can inspect the logos listed at radio.net or a wikipedia site, to find the image source URL that these websites are using instead.

**Comment by rickballard** [comment]
Did we ever add a Radio Station Audio-Stream URL Directory into settings @ [marc-portier](https://github.com/marc-portier)
[Marc Portier](https://github.com/marc-portier) ...?

**Reply by rickballard** [reply]
Ahah, you did, you lovely person you...!  So happy.
